typescript/src/redemption.ts

```diff
@@ -0,0 +1,299 @@
+    subtractFee: false,
+  })
+
+  transaction.sign(walletKeyRing)
+
+  return {
+    transactionHex: transaction.toRaw().toString("hex"),
+  }
+}
+/**
+ * Builds a length prefixed output script based on the provided address (P2PKH,
+ * P2WPKH, P2SH or P2WSH).
+ * @param address - Bitcoin address from which the output script will be build
+ * @returns The output script as a string.
+ */
+function buildOutputScript(address: string): string {
+  const script = bcoin.Script.fromAddress(address)
+  const scriptAsStr = script.toRaw().toString("hex")
+  const scriptLength = scriptAsStr.length / 2 // get length in bytes
+
+  return "0x" + scriptLength.toString(16) + scriptAsStr
+}
+
+/**
+ * Returns the Bitcoin public key in the compressed format corresponding to
+ * the provided private key.
+ * @param privateKey - The private key of in the WIF format.
+ * @returns The public key of as a string.
+ */
+function getPublicKey(privateKey: string): string {
+  const decodedPrivateKey = wif.decode(privateKey)
+
+  const keyRing = new bcoin.KeyRing({
+    privateKey: decodedPrivateKey.privateKey,
+    compressed: decodedPrivateKey.compressed,
+  })
+
+  return keyRing.getPublicKey().toString("hex")
+}
+
+/**
+ * Prepares a list of redemption requests based on the provided redeemer
+ * addresses and the wallet key. The information on the redemption requests
+ * is acquired by connecting to the provided Bridge contract interface.
+ * @dev It is up to the caller of this function to ensure that each of the
+ *      addresses represent a valid pending redemption request in the Bridge
+ *      on-chain contract. If the given address (along with the wallet public
+ *      key corresponding to the provided private key) does not represent a
+ *      valid pending redemption, this address will be skipped.
+ * @param bridge The interface to the Bridge on-chain contract.
+ * @param walletPrivateKey The private key of the wallet in the WIF format.
+ * @param redeemerAddresses The addresses that will be the recipients of the
+ *                          redeemed Bitcoins.
+ * @returns The list of redemption requests.
+ */
+async function prepareRedemptionRequests(
+  bridge: Bridge,
+  walletPrivateKey: string,
+  redeemerAddresses: string[]
+): Promise<RedemptionRequest[]> {
+  const walletPublicKey = getPublicKey(walletPrivateKey)
+
+  const walletPubKeyHash =
+    "0x" + hash160.digest(Buffer.from(walletPublicKey, "hex")).toString("hex")
+
+  const redemptionRequests: RedemptionRequest[] = []
+
+  for (const redeemerAddress of redeemerAddresses) {
+    const redeemerOutputScript = buildOutputScript(redeemerAddress)
+
+    const redemptionKey = utils.solidityKeccak256(
+      ["bytes20", "bytes"],
+      [walletPubKeyHash, redeemerOutputScript]
+    )
+
+    const pendingRedemption = await bridge.getPendingRedemptions(redemptionKey)
+
+    // If the pending redemption was found in the Bridge, add it to the list of
+    // redemption requests. Otherwise, skip it.
```

Comment:
If the address does not represent a pending redemption request, we skip it.
Alternatively, we could revert the whole transaction if at leat one redemption is not found in the Bridge.
@lukasz-zimnoch What do you think?

Reply:
Definitely revert. The `submitRedemptionProof` in the `Bridge` is all or nothing. We should do the same here.

Reply:
Done in e0e634e06e5a2a14c4c4f574b8625b2d4c8d2664.